Document our config files and structure

petr-balogh opened the pull request.

ebenahar commented on conf/README.md:
```diff
@@ -0,0 +1,84 @@
+In this directory we will be storing all the configuration for clusters and
+OCSCI config files.
+
+During the execution we are via
```
This sentence is a bit unclear. Change to something like:
During the execution, via [ocsci pytest plugin](../pytest_customization/ocscilib.py), we're loading all the...

petr-balogh replied:
done

ebenahar commented on conf/README.md:
```diff
@@ -0,0 +1,84 @@
+rendering variables which you can reuse among different configs from DEFAULT
+and different sections.
+
+This is giving use the ability to easily reusing variables from context of the
```
This gives the ability to use

petr-balogh replied:
done

ebenahar commented on conf/README.md:
```diff
@@ -0,0 +1,84 @@
+    templates whic allow us to add new values just in config and we don't have
+    to also update the code on all places to take specific variable.
+
+If you defining cluster config you have to put all the data under `ENV_DATA`
```
If you define

petr-balogh replied:
done

ebenahar commented on conf/README.md:
```diff
@@ -0,0 +1,84 @@
+    to also update the code on all places to take specific variable.
+
+If you defining cluster config you have to put all the data under `ENV_DATA`
+section like in mentioned example. Actually you can add another section as
```
in the mentioned example

petr-balogh replied:
done

ebenahar commented on conf/README.md:
```diff
@@ -0,0 +1,84 @@
+
+### Sections in our config
+
+All of the sections below except `DEFAULTS` section  will be available from
```
that will be available

petr-balogh replied:
done

ebenahar commented on conf/README.md:
```diff
@@ -0,0 +1,84 @@
+    `ENV_DATA` section to be able to reuse in templates.
+* REPORTING - reporting related config. (Do not store secret data in the
+    repository!).
+* ENV_DATA - this section is meant to be overwritten by onw cluster config
```
one

petr-balogh replied:
done

ebenahar commented on conf/README.md:
```diff
@@ -0,0 +1,84 @@
+
+### Sections in our config
+
+All of the sections below except `DEFAULTS` section that will be available
```
Still need to be rephrased. Maybe something like:

All of the sections below, except for 'DEFAULTS' one, will be available from..

vasukulkarni commented:
Waiting for the explanation in our regular meeting because I am not getting how the defaults is working.

shylesh commented:
@petr-balogh 
Do you want to add something related to suites organization as well here ?

Thanks,
Shylesh

mbukatov commented on conf/README.md:
```diff
@@ -0,0 +1,84 @@
+
+If you define cluster config you have to put all the data under `ENV_DATA`
+section like in the mentioned example. Actually you can add another section as
+well but we recommend int cluster config define just `ENV_DATA` section or one
```
```suggestion
well but in a cluster config file we recommend to define just `ENV_DATA` section or one
```

petr-balogh replied:
done

mbukatov commented on conf/README.md:
```diff
@@ -0,0 +1,84 @@
+    [../templates/ocs-deployment/] and you can see the variables in templates.
+    In the deploment we are using ENV_DATA section as data for rendering the
+    templates whic allow us to add new values just in config and we don't have
+    to also update the code on all places to take specific variable.
```
Actually, I would suggest to put this into the paragraph below the list, which describes ENV_DATA as well, and together it makes the point better.

mbukatov commented on conf/README.md:
```diff
@@ -0,0 +1,84 @@
+[ocsci folder](./ocsci/). You can see we have defined here:
+
+* [default_config.yaml](ocsci/default_config.yaml) - this is the default
+    config file for OCS CI whih is loaded automatically (don't pass it to
```
```suggestion
    config file for OCS CI which is loaded automatically (don't pass it to
```

petr-balogh replied:
done

mbukatov commented on conf/README.md:
```diff
@@ -0,0 +1,84 @@
+more additional section which is `DEFAULTS` for propagation those defaults
+among all configs as well.
+
+## OCS CI Config
```
Now I think that it would be better to move this section to the top, making it the 1st one.  This way, you will explain that we have OCS CI config, defaults and sections in our confing. Then it is easier to describe the cluster config (where you talk about the defaults and some sections anyway, but without broader context).

petr-balogh replied:
moved

mbukatov commented on conf/README.md:
```diff
@@ -0,0 +1,84 @@
+* ENV_DATA - this section is meant to be overwritten by own cluster config
+    file, but can be overwritten also here (But cluster config has higher
+    priority).
+
```
It would be nice to have there an example how to read particular config value in python code (based either on the default config or some simple example).

mbukatov commented on conf/README.md:
```diff
@@ -0,0 +1,84 @@
+
+1) **CLI args** - sometime we can pass some variables by CLI parameters, in
+    this case those arguments should overwrite everything and have the highest
+    priority.
```
Do I understand it right that so far, there is only cluster name which works this way?

petr-balogh replied:
Yep, currently only this one but in future we may have more.

vasukulkarni commented on conf/README.md:
```diff
@@ -0,0 +1,123 @@
+files.
+
+During the execution, via
+[ocsci pytest plugin](../pytest_customization/ocscilib.py), we are loading all
```
this statement needs to be rewritten, probably " During the execution we are loading the config file with different sections (Defaults, RUN etc) and using the values defined in the config file to render yaml with jinja2 templates " ?

vasukulkarni commented on conf/README.md:
```diff
@@ -0,0 +1,123 @@
+This gives us the ability to easily reuse variables from context of the
+yaml file itself.
+
+Cause you can overwrite data just in specific section like for ENV_DATA
```
I would cut the paragraph as its implied by config file context.

vasukulkarni commented on conf/README.md:
```diff
@@ -0,0 +1,123 @@
+
+* [default_config.yaml](ocsci/default_config.yaml) - this is the default
+    config file for OCS CI which is loaded automatically (don't pass it to
+    --ocsci-conf). It contains multiple sections which you can overwrite in
```
Line starting with "It contains multiple sections" should be rewritten,

"It contains multiple sections which user can overwrite in custom file and provide as an option to --oscci-conf"

vasukulkarni commented on conf/README.md:
```diff
@@ -0,0 +1,123 @@
+## Cluster config
+
+If you would like to overwrite cluster default data you can create file
+similar to [this example](./examples/ocs_basic_install.yml). You can see we
```
s/You can see we can overwrite default data like/example shown overwrites below ENV data/

vasukulkarni commented on conf/README.md:
```diff
@@ -0,0 +1,123 @@
+
+### Sections in our configs
+
+All of the sections below, except for `DEFAULTS` one, will be available from
```
s/sections below/below sections/ ; s/one// ;

petr-balogh replied:
this was suggested sentence from Elad from different review ;) , but updated in next commit.

vasukulkarni commented on conf/README.md:
```diff
@@ -0,0 +1,123 @@
+
+Default values which are used among different configs and sections.
+
+All **capital letters** defined variables are exposed to
```
why just CAPS, why not all? you probably want to simple the next 7 lines and just say "Defaults will be exported via defaults.py"

petr-balogh replied:
In the defaults we would like to have mainly constants. So if we follow pep8 convention constants are with CAPS. I can put there everything, cause I solved the issue with loading  this config which I will send in new commit.

vasukulkarni commented on conf/README.md:
```diff
@@ -0,0 +1,123 @@
+#### DEPLOYMENT
+
+Deployment related parameters. In deplyment we are using some config from this
+section but if its ENV specific we are using also `ENV_DATA` section to be
```
ENV mix is bit confusing, can be just moved to ENV section

vasukulkarni commented on conf/README.md:
```diff
@@ -0,0 +1,123 @@
+
+# From you code you can access those data like
+
+# Taking data from ENV_DATA will alsways using actuall cluster_namespace
```
Can you repharse this, looks like late night edition :-)

petr-balogh replied:
done in next commit I will send

vasukulkarni commented on utility/utils.py:
```diff
@@ -725,20 +724,24 @@ def destroy_cluster(cluster_path):
-def get_openshift_installer(version=defaults.INSTALLER_VERSION):
+def get_openshift_installer(
+    version=DEPLOYMENT['installer_version'],
+    bin_dir=RUN['bin_dir'],
```
this should also be DEPLOYMENT var ?

dahorak replied:
IMHO, I would put all three variables: `installer_version`, `client_version` and `bin_dir` to the DEPLOYMENT section, as it is related to "deployment" - no matter if it is deployment of installer or client...

petr-balogh replied:
I am thinking about RUN section which is specific for run itself, we can run without deployment at all, on deployed env by RHHI.next for example.  OC client will be used in run itself and we will need to download it on the start of run anyway to make sure we are using version from config, or have session fixture which make sure it's installed/downloaded. the bin_dir will be probably common for whole run and we can put there more binaries files in the future so everything in this dir will be available in PATH. That's why I decided for RUN section for those.  Installer version is just for deployment, nothing else, that's why in DEPLOYMENT section.  @dahorak @vasukulkarni  are you OK with this?

dahorak replied:
Ok

vasukulkarni commented on tests/test_pv_creation.py:
```diff
@@ -21,7 +21,7 @@
 
 OCP = ocp.OCP(
-    kind='PersistentVolume', namespace=defaults.ROOK_CLUSTER_NAMESPACE
+    kind='PersistentVolume', namespace=ENV_DATA['cluster_namespace']
```
since the defaults is populated from yaml, I think leaving this as is makes it more readable?

vasukulkarni replied:
Ignore this one please, Its ENV here and not defaults.

petr-balogh commented on conf/ocsci/README.md:
```diff
@@ -0,0 +1,6 @@
+In this directory we will have common config files which we will use in our
+automation for example for different platforms like AWS or VmWare and so on.
+
+See the documentation from [conf dir][../README.md] for more details!
```
Done

petr-balogh commented on conf/README.md:
```diff
@@ -0,0 +1,84 @@
+# Config directory
+
+In this directory we will be storing all the configuration for clusters and
```
done

petr-balogh commented on conf/README.md:
```diff
@@ -0,0 +1,84 @@
+templates rendering variables which you can reuse among different configs from
+DEFAULT and different sections.
+
+This gives us the ability to easily reusing variables from context of the
```
done

petr-balogh commented on conf/README.md:
```diff
@@ -0,0 +1,84 @@
+* `worker_replicas` - number of replicas of worker.
+* `rook_image` - which rook image can be used.
+* and many more - look at the deployment templates under
+    [../templates/ocs-deployment/] and you can see the variables in templates.
```
done

dahorak commented on utility/utils.py:
```diff
@@ -756,24 +759,29 @@ def get_openshift_installer(version=defaults.INSTALLER_VERSION):
-def get_openshift_client(version=defaults.CLIENT_VERSION):
+def get_openshift_client(
+    version=RUN['client_version'],
+    bin_dir=RUN['bin_dir'],
```
As mentioned above, I would put both `client_version` and `bin_dir` variables also to the DEPLOYMENT section... but consider it just as my POV...

vasukulkarni replied:
Agreeing to previous similar comment, so resolving this.

petr-balogh commented:
@vasukulkarni  can you please take a look? I am on PTO next week, so just running verification of this.  If there are still some unclear sentences or typos I guess it can be done in some next PR also by someone else in next week.  I had to added new wrapper for pytest which allow me to bootstrap our config before the pytest starting loading any other module.

vasukulkarni commented:
@petr-balogh this looks good to me, some minor nits probably we can take up as follow up PR and I can submit that.

vasukulkarni commented:
@petr-balogh I dont think this needs to be squashed and merged, Correct?

vasukulkarni commented:
I can see they are independent.

petr-balogh commented:
> @petr-balogh I dont think this needs to be squashed and merged, Correct?

no, I squashed some commits last time I submitted cause I needed to rebase anyway. Verification job is still running and deploying cluster.

A lot of PRs probably will need to be rebased on this change as I am also deleting functionality to directly import from ocsci all those markers and base test classes. Moved this to ocsci/testlib.py  so ppl will import from it now.

petr-balogh commented:
@vasukulkarni was quick, thanks, I will rebase the patch for destroy on top of this one and we can merge as well.

petr-balogh commented:
A bit late with verification deployment test passed but I see some warnings which I didn't hit before:
Verification of deployment:
```bash
(.venv) [jenkins@qe-ocs4-slave-prod-3 ocs-ci]$ python -m run_ocsci -m deployment --cluster-name pbalogh-config --clus
ter-path /home/jenkins/config-test-cluster ./tests
The kubeconfig file /home/jenkins/config-test-cluster/auth/kubeconfig doesn't exist!
================================================ test session starts ================================================
platform linux -- Python 3.6.6, pytest-4.5.0, py-1.8.0, pluggy-0.12.0 -- /home/jenkins/ocs-ci/.venv/bin/python
cachedir: .pytest_cache
rootdir: /home/jenkins/ocs-ci, inifile: pytest.ini
plugins: ocs-ci-0.1
collected 5 items / 4 deselected / 1 selected                                                                       

tests/ecosystem/deployment/test_ocs_basic_install.py::TestDeployment::test_deployment
--------------------------------------------------- live log call ---------------------------------------------------
22:45:09 - MainThread - tests.ecosystem.deployment.test_ocs_basic_install - INFO - Running OCS basic installation
22:45:09 - MainThread - oc.openshift_ops - INFO - Testing access to cluster with /home/jenkins/config-test-cluster/au
th/kubeconfig
22:45:09 - MainThread - oc.openshift_ops - WARNING - The kubeconfig file /home/jenkins/config-test-cluster/auth/kubec
onfig doesn't exist!
22:45:09 - MainThread - tests.ecosystem.deployment.test_ocs_basic_install - INFO - Generating install-config
22:45:09 - MainThread - utility.utils - INFO - Executing command: mkdir -p /home/jenkins/config-test-cluster
22:45:09 - MainThread - tests.ecosystem.deployment.test_ocs_basic_install - INFO - Install config:
.... output omitted ....
23:21:06 - MainThread - utility.utils - INFO - Executing command: oc -n openshift-storage exec rook-ceph-tools-76c7d559b6-4tlzq ceph health
23:21:07 - MainThread - tests.ecosystem.deployment.test_ocs_basic_install - INFO - HEALTH_OK, install successful.
PASSED                                                                                                        [100%]

================================================= warnings summary ==================================================
.venv/lib64/python3.6/site-packages/_pytest/config/__init__.py:781
.venv/lib64/python3.6/site-packages/_pytest/config/__init__.py:781
  /home/jenkins/ocs-ci/.venv/lib64/python3.6/site-packages/_pytest/config/__init__.py:781: PytestAssertRewriteWarning: Module already imported so cannot be rewritten: ocs
    self._mark_plugins_for_rewrite(hook)

.venv/lib64/python3.6/site-packages/_pytest/config/__init__.py:781
.venv/lib64/python3.6/site-packages/_pytest/config/__init__.py:781
  /home/jenkins/ocs-ci/.venv/lib64/python3.6/site-packages/_pytest/config/__init__.py:781: PytestAssertRewriteWarning: Module already imported so cannot be rewritten: ocsci
    self._mark_plugins_for_rewrite(hook)

-- Docs: https://docs.pytest.org/en/latest/warnings.html
=============================== 1 passed, 4 deselected, 4 warnings in 2157.80 seconds ===============================
````

Will try to take a look at the weekend, not sure what causing those warnings.

vasukulkarni commented:
> > @petr-balogh I dont think this needs to be squashed and merged, Correct?
> 
> no, I squashed some commits last time I submitted cause I needed to rebase anyway. Verification job is still running and deploying cluster.
> 
> A lot of PRs probably will need to be rebased on this change as I am also deleting functionality to directly import from ocsci all those markers and base test classes. Moved this to ocsci/testlib.py so ppl will import from it now.

like this? https://github.com/red-hat-storage/ocs-ci/blob/master/tests/test_add_mds_to_cluster.py#L13

If there are more we can quickly fix them.

petr-balogh commented:
@vasukulkarni yes, I tried to fix what was in master. But ppl will need to fix in their PRs ;)

Thanks, have to go sleep :)

vasukulkarni commented:
@petr-balogh np thanks :-)
